### ISSUE TITLE
Property name "constructor" is not allowed in 'Model' data

### ISSUE THREAD

**Issue by masterclock** [strongloop/loopback-datasource-juggler#1261]
### Bug or feature request

- [ ] Bug
- [x ] Feature request

### Description of feature (or steps to reproduce if bug)
I'm using loopback with Apollostack's graphql server.
`ModelBaseClass.prototype._initProperties` function requires input `data` be `undefined` or `data.constructor` be a function.
I hope and it's reasonable that it accepts with `data.constructor` being 'undefined', as graphql's resolver function 
passes data with no constructor.

### Expected result
accepts with `data.constructor` being 'undefined'.

### Actual result (if bug)
Loopback throws error

### Additional information (Node.js version, LoopBack version, etc)
loopback-datasource-juggler: 3.2.0

**Comment by langxiong** [comment]
I got `data = null`, 
then `data.constructor` would cause the program exit with 
```
TypeError: Cannot read property 'constructor' of null
    at ModelConstructor.ModelBaseClass._initProperties (c:\Users\x5831\crawler\tc-server\node_modules\loopback-datasource-juggler\lib\model.js:76:19)

```

## Additional information (Node.js version, LoopBack version, etc)

loopback-datasource-juggler: 3.2.0

**Comment by jannyHou** [comment]
@masterclock @langxiong "constructor" is a reserved word in loopback,  please check doc to see more details: http://loopback.io/doc/en/lb3/Valid-names-in-LoopBack.html#general-rules details 

And define a property called "constructor" would cause error in loopback code, that's why we decide to throw error when model has it as property. There is a discussion in issue https://github.com/strongloop/loopback-datasource-juggler/issues/762 explains how it breaks the code.

**Comment by jannyHou** [comment]
UPDATE:
@masterclock If you still want to make it work in loopback, can you elaborate more about the step to reproduce your error? Maybe fork https://github.com/strongloop/loopback-sandbox and replace with the code how you use https://github.com/apollographql/graphql-server in loopback, we will investigate if it's a feasible feature. Thanks!

**Comment by masterclock** [comment]
I'm not using `constructor` as a property name, but creating model instance with some data whose `constructor` is `null` or `undefined`.

say: 
```
    const model = app.models.TestModel;
    let obj = Object.create(null); // obj's constructor === undefined
    const instance = model(obj); // fail!!
```

you may see repo as a test:
`https://github.com/masterclock/test-loopback-no-constructor/tree/master`

**Comment by Tallyb** [comment]
I am having the same problem in my loopback-graphql package. 
You can run the mutation tests here: https://github.com/Tallyb/loopback-graphql
The problem occurs on resolvers, line 36.

**Comment by Tallyb** [comment]
@masterclock  Here is a workaround I used (not a very pleasant one, but it works): 
JSON.parse(JSON.stringify(obj)));

**Comment by jannyHou** [comment]
Thank you @masterclock I appreciate your example repo 
And @Tallyb your repo looks cool :) 
We will look into the problem soon.

**Comment by thaerlabs** [comment]
looking at the code:
```javascript
if (typeof data !== 'undefined' &&
    typeof (data.constructor) !== 'function') {
    throw new Error(g.f('Property name "{{constructor}}" is not allowed in %s data', ctor.modelName));
}
```

What if we pass an object like:
```javascript
data = {
  name: "this is my name"
}
```

Plain objects will make the condition be **falsey** as there's no `constructor` prop.

My approach is to check before, if the `data` object has a `constructor` property in the first place.


@jannyHou I opened a PR for this, can you please check it.
https://github.com/strongloop/loopback-datasource-juggler/pull/1284

**Comment by jonathanfoster** [comment]
I'm upgrading from LoopBack 2x to 3x and I ran into this same problem. The `typeof(data.constructor) !== 'function'` returns true because data.constructor is undefined. I tried @thaerlabs' fix in https://github.com/strongloop/loopback-datasource-juggler/pull/1284 and it solved the issue for me.

**Comment by ssh24** [comment]
Closing this. Issue fixed in: #1284
